File: examples/nodejs/checker_queries.ts
import { TsgoApiClient } from "@corsa-bind/napi";
import type { ProjectResponse, UpdateSnapshotResponse } from "@corsa-bind/napi";

import { assertExists, isMain, mockBinary, workspaceRoot } from "../shared.ts";

interface SymbolResponse {
  id: string;
  name: string;
  valueDeclaration?: string;
}

interface TypeResponse {
  id: string;
  symbol?: string;
  texts: string[];
}

interface SignatureResponse {
  id: string;
  parameters: string[];
}

interface TypePredicateResponse {
  parameterName?: string;
  type?: TypeResponse;
}

interface IndexInfo {
  keyType: TypeResponse;
  valueType: TypeResponse;
  isReadonly: boolean;
}

function requireProject(snapshot: UpdateSnapshotResponse): ProjectResponse {
  const project = snapshot.projects[0];
  if (!project) {
    throw new Error("checker queries example did not return a project");
  }
  return project;
}

export function runCheckerQueriesExample() {
  assertExists(
    mockBinary,
    "mock tsgo binary",
    "run `vp run -w build_mock` or `vp run -w build` first",
  );

  const client = TsgoApiClient.spawn({
    executable: mockBinary,
    cwd: workspaceRoot,
    mode: "jsonrpc",
  });

  let snapshotHandle: string | undefined;

  try {
    const init = client.initialize();
    const snapshot = client.updateSnapshot({
      openProject: "/workspace/tsconfig.json",
    });
    snapshotHandle = snapshot.snapshot;
    const project = requireProject(snapshot);

    const symbol = client.callJson<SymbolResponse | null>("getSymbolAtPosition", {
      snapshot: snapshot.snapshot,
      project: project.id,
      file: "/workspace/src/index.ts",
      position: 1,
    });
    if (!symbol) {
      throw new Error("checker queries example did not resolve a symbol");
    }

    const apparentType = client.callJson<TypeResponse | null>("getTypeOfSymbol", {
      snapshot: snapshot.snapshot,
      project: project.id,
      symbol: symbol.id,
    });
    if (!apparentType) {
      throw new Error("checker queries example did not resolve an apparent type");
    }

    const declaredType = client.callJson<TypeResponse | null>("getDeclaredTypeOfSymbol", {
      snapshot: snapshot.snapshot,
      project: project.id,
      symbol: symbol.id,
    });
    const typeAtPosition = client.callJson<TypeResponse | null>("getTypeAtPosition", {
      snapshot: snapshot.snapshot,
      project: project.id,
      file: "/workspace/src/index.ts",
      position: 1,
    });
    const batchTypes =
      client.callJson<Array<TypeResponse | null> | null>("getTypesAtPositions", {
        snapshot: snapshot.snapshot,
        project: project.id,
        file: "/workspace/src/index.ts",
        positions: [1, 2],
      }) ?? [];
    const signatures = client.callJson<SignatureResponse[]>("getSignaturesOfType", {
      snapshot: snapshot.snapshot,
      project: project.id,
      type: apparentType.id,
      kind: 0,
    });
    const primarySignature = signatures[0];
    if (!primarySignature) {
      throw new Error("checker queries example did not return a signature");
    }

    const returnType = client.callJson<TypeResponse | null>("getReturnTypeOfSignature", {
      snapshot: snapshot.snapshot,
      project: project.id,
      signature: primarySignature.id,
    });
    const predicate = client.callJson<TypePredicateResponse | null>("getTypePredicateOfSignature", {
      snapshot: snapshot.snapshot,
      project: project.id,
      signature: primarySignature.id,
    });
    const properties =
      client.callJson<SymbolResponse[] | null>("getPropertiesOfType", {
        snapshot: snapshot.snapshot,
        project: project.id,
        type: apparentType.id,
      }) ?? [];
    const indexInfos =
      client.callJson<IndexInfo[] | null>("getIndexInfosOfType", {
        snapshot: snapshot.snapshot,
        project: project.id,
        type: apparentType.id,
      }) ?? [];
    const resolved = client.callJson<SymbolResponse | null>("resolveName", {
      snapshot: snapshot.snapshot,
      project: project.id,
      name: "value",
      meaning: 2,
      file: "/workspace/src/index.ts",
      position: 1,
    });
    const exported = client.callJson<SymbolResponse>("getExportSymbolOfSymbol", {
      snapshot: snapshot.snapshot,
      symbol: symbol.id,
    });

    return {
      currentDirectory: init.currentDirectory,
      projectId: project.id,
      symbol: {
        id: symbol.id,
        name: symbol.name,
        valueDeclaration: symbol.valueDeclaration ?? null,
      },
      types: {
        apparent: apparentType.texts[0] ?? null,
        apparentId: apparentType.id,
        declared: declaredType?.texts[0] ?? null,
        position: typeAtPosition?.texts[0] ?? null,
        batchIds: batchTypes.map((item) => item?.id ?? null),
        returnType: returnType?.texts[0] ?? null,
        predicateType: predicate?.type?.texts[0] ?? null,
      },
      signatures: signatures.map((signature) => ({
        id: signature.id,
        parameterCount: signature.parameters.length,
      })),
      properties: properties.map((property) => property.name),
      indexInfos: indexInfos.map((info) => ({
        keyType: info.keyType.texts[0] ?? null,
        valueType: info.valueType.texts[0] ?? null,
        isReadonly: info.isReadonly,
      })),
      resolvedName: resolved?.name ?? null,
      exportedName: exported.name,
    };
  } finally {
    if (snapshotHandle) {
      client.releaseHandle(snapshotHandle);
    }
    client.close();
  }
}

if (isMain(import.meta.url)) {
  console.log(JSON.stringify(runCheckerQueriesExample(), null, 2));
}

File: examples/rust/checker_queries.rs
mod support;

use corsa::{
    TsgoError,
    api::{ApiClient, ApiMode, DocumentPosition, UpdateSnapshotParams},
    runtime::block_on,
};
use serde_json::json;

fn main() -> Result<(), corsa::TsgoError> {
    let result = block_on(async {
        let client = ApiClient::spawn(support::mock_api_config(
            "checker_queries",
            ApiMode::AsyncJsonRpcStdio,
        )?)
        .await?;
        let snapshot = client
            .update_snapshot(UpdateSnapshotParams {
                open_project: Some("/workspace/tsconfig.json".into()),
                file_changes: None,
                overlay_changes: None,
            })
            .await?;
        let project = snapshot.projects.first().ok_or_else(|| {
            TsgoError::Protocol("checker queries example did not return a project".into())
        })?;
        let symbol = client
            .get_symbol_at_position(
                snapshot.handle.clone(),
                project.id.clone(),
                "/workspace/src/index.ts",
                1,
            )
            .await?
            .ok_or_else(|| {
                TsgoError::Protocol("checker queries example did not resolve a symbol".into())
            })?;
        let declaration = symbol.value_declaration.clone().ok_or_else(|| {
            TsgoError::Protocol("checker queries example did not return a declaration".into())
        })?;
        let parsed = declaration.parse()?;
        let type_at_position = client
            .get_type_at_position(
                snapshot.handle.clone(),
                project.id.clone(),
                "/workspace/src/index.ts",
                1,
            )
            .await?
            .ok_or_else(|| {
                TsgoError::Protocol("checker queries example did not resolve a type".into())
            })?;
        let type_at_location = client
            .get_type_at_location(
                snapshot.handle.clone(),
                project.id.clone(),
                declaration.clone(),
            )
            .await?
            .ok_or_else(|| {
                TsgoError::Protocol(
                    "checker queries example did not resolve a location type".into(),
                )
            })?;
        let declared_type = client
            .get_declared_type_of_symbol(
                snapshot.handle.clone(),
                project.id.clone(),
                symbol.id.clone(),
            )
            .await?
            .ok_or_else(|| {
                TsgoError::Protocol(
                    "checker queries example did not resolve a declared type".into(),
                )
            })?;
        let signatures = client
            .get_signatures_of_type(
                snapshot.handle.clone(),
                project.id.clone(),
                type_at_position.id.clone(),
                0,
            )
            .await?;
        let signature = signatures.first().ok_or_else(|| {
            TsgoError::Protocol("checker queries example did not return a signature".into())
        })?;
        let return_type = client
            .get_return_type_of_signature(
                snapshot.handle.clone(),
                project.id.clone(),
                signature.id.clone(),
            )
            .await?
            .ok_or_else(|| {
                TsgoError::Protocol("checker queries example did not resolve a return type".into())
            })?;
        let predicate = client
            .get_type_predicate_of_signature(
                snapshot.handle.clone(),
                project.id.clone(),
                signature.id.clone(),
            )
            .await?;
        let properties = client
            .get_properties_of_type(
                snapshot.handle.clone(),
                project.id.clone(),
                type_at_position.id.clone(),
            )
            .await?;
        let index_infos = client
            .get_index_infos_of_type(
                snapshot.handle.clone(),
                project.id.clone(),
                type_at_position.id.clone(),
            )
            .await?;
        let batch_symbols = client
            .get_symbols_at_positions(
                snapshot.handle.clone(),
                project.id.clone(),
                "/workspace/src/index.ts",
                vec![1, 2],
            )
            .await?;
        let batch_types = client
            .get_types_at_positions(
                snapshot.handle.clone(),
                project.id.clone(),
                "/workspace/src/index.ts",
                vec![1, 2],
            )
            .await?;
        let resolved = client
            .resolve_name_at_position(
                snapshot.handle.clone(),
                project.id.clone(),
                "value",
                2,
                DocumentPosition {
                    document: "/workspace/src/index.ts".into(),
                    position: 1,
                },
                None,
            )
            .await?;
        let symbol_of_type = client
            .get_symbol_of_type(snapshot.handle.clone(), type_at_position.id.clone())
            .await?;
        let exported = client
            .get_export_symbol_of_symbol(snapshot.handle.clone(), symbol.id.clone())
            .await?;

        let result = json!({
            "projectId": project.id,
            "symbol": {
                "id": symbol.id,
                "name": symbol.name,
                "declaration": {
                    "raw": declaration,
                    "kind": parsed.kind,
                    "path": parsed.path,
                    "pos": parsed.pos,
                    "end": parsed.end,
                },
            },
            "types": {
                "atPosition": type_at_position.texts.first(),
                "atLocation": type_at_location.texts.first(),
                "declared": declared_type.texts.first(),
                "returnType": return_type.texts.first(),
                "predicateType": predicate
                    .as_ref()
                    .and_then(|item| item.r#type.as_ref())
                    .and_then(|item| item.texts.first()),
            },
            "batch": {
                "symbolNames": batch_symbols
                    .iter()
                    .map(|item| item.as_ref().map(|item| item.name.clone()))
                    .collect::<Vec<_>>(),
                "typeIds": batch_types
                    .iter()
                    .map(|item| item.as_ref().map(|item| item.id.clone()))
                    .collect::<Vec<_>>(),
            },
            "signature": {
                "id": signature.id,
                "parameterCount": signature.parameters.len(),
            },
            "properties": properties.iter().map(|item| item.name.clone()).collect::<Vec<_>>(),
            "indexInfos": index_infos
                .iter()
                .map(|item| json!({
                    "keyType": item.key_type.texts.first(),
                    "valueType": item.value_type.texts.first(),
                    "isReadonly": item.is_readonly,
                }))
                .collect::<Vec<_>>(),
            "resolvedName": resolved.as_ref().map(|item| item.name.clone()),
            "symbolOfType": symbol_of_type.as_ref().map(|item| item.name.clone()),
            "exportedName": exported.name,
        });
        snapshot.release().await?;
        client.close().await?;
        Ok::<_, corsa::TsgoError>(result)
    })?;

    support::print_json(result);
    Ok(())
}
